airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py

```diff
@@ -0,0 +1,153 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import configparser
+import io
+import uuid
+from enum import Enum, auto
+from typing import Dict, Optional, Tuple
+
+import tomli
+import tomli_w
+from dagger import Container, Directory
+from pipelines.airbyte_ci.poetry.publish.context import PyPIPublishContext
+from pipelines.consts import PYPROJECT_TOML_FILE_PATH, SETUP_PY_FILE_PATH
+from pipelines.dagger.actions.python.poetry import with_poetry
+from pipelines.helpers.utils import sh_dash_c
+from pipelines.models.steps import Step, StepResult
+
+
+class PackageType(Enum):
+    POETRY = auto()
+    PIP = auto()
+    NONE = auto()
+
+
+class PublishToPyPI(Step):
+    context: PyPIPublishContext
+    title = "Publish package to PyPI"
+
+    def _get_base_container(self) -> Container:
+        return with_poetry(self.context)
+
+    async def _get_package_metadata_from_pyproject_toml(self, dir_to_publish: Directory) -> Optional[Tuple[str, str]]:
+        pyproject_toml = dir_to_publish.file(PYPROJECT_TOML_FILE_PATH)
+        pyproject_toml_content = await pyproject_toml.contents()
+        contents = tomli.loads(pyproject_toml_content)
+        if (
+            "tool" not in contents
+            or "poetry" not in contents["tool"]
+            or "name" not in contents["tool"]["poetry"]
+            or "version" not in contents["tool"]["poetry"]
+        ):
+            return None
+
+        return (contents["tool"]["poetry"]["name"], contents["tool"]["poetry"]["version"])
+
+    async def _get_package_type(self, dir_to_publish: Directory) -> PackageType:
+        files = await dir_to_publish.entries()
+        has_pyproject_toml = PYPROJECT_TOML_FILE_PATH in files
+        has_setup_py = SETUP_PY_FILE_PATH in files
+        if has_pyproject_toml:
+            return PackageType.POETRY
+        elif has_setup_py:
+            return PackageType.PIP
+        else:
+            return PackageType.NONE
+
+    async def _run(self) -> StepResult:
```

Comment:
Readability improvement suggestion:
Would it be possible to slightly rework this method so that when reading it we don't have to follow the `if poetry` or `if pip`? The switch / fork could happen in sub methods.